# 4.7 Deriving New Variables

### Import Libraries: Pandas, Numpy, Os

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

### Create path to 'Prepared data'

In [2]:
# Create path to Prepared data folder
path_prepared = r'C:\Users\danie\Desktop\career Foundry\31-03-2023 Instacart Basket Analysis\02 Data\Prepared data'

### Import DataFrame, Pickle file (.pkl)

In [3]:
# Import .pkl
df_ords_prods = pd.read_pickle(os.path.join(path_prepared, 'orders_products_merged.pkl'))

### Check columns & shape

In [4]:
# Check dataframe with '.head()'
df_ords_prods.head()

Unnamed: 0_x  order_id  user_id  order_number orders_day_of_the_week  \
0             0   2539329        1             1                 Monday   
1             1   2398795        1             2                Tuesday   
2             2    473747        1             3                Tuesday   
3             3   2254736        1             4              Wednesday   
4             4    431534        1             5              Wednesday   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge  Unnamed: 0_y product_name  aisle_id  department_id  \
0          0   both           195         Soda        77              7   
1          1   both           195         Soda        77              7   
2          1   both           195         Soda        77              7   
3          1   both           195         Soda        77              7   
4          1   both           195         Soda        77              7   

   prices  
0     9.0  
1     9.0  
2     9.0  
3     9.0  
4     9.0

In [5]:
# Check dataframe shape
df_ords_prods.shape

(32404859, 16)

***The shape of the file 'orders_products_merged.pkl' imported match with the shape of the file previously exported in the last exercise.***

### Create subset with first 1Million rows

In [6]:
# Subset df with the first million rows 
df = df_ords_prods[:1000000]

In [7]:
df.shape

(1000000, 16)

### Run merge tests

#.merge test with '.., how = outer, left, right, inner join' conditions has been checked before procceding to the next step.

### Create new variable using 'def' with conditional function 'if, elif, else'.

In [8]:
# using 'def' with conditional function 'if, elif, else'.
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range'
    else: 
        return 'Not enough data'

In [9]:
#Creating new variable 'price_range'
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\danie\AppData\Local\Temp\ipykernel_4460\3078710985.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [10]:
# .value_counts() count of entries with the same values in a column
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [11]:
# Checking Max value in 'prices'
df['prices'].max()

14.8

### If-Statements with the loc() Function

In [12]:
# Using .loc() Function to add values in a new column
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\danie\AppData\Local\Temp\ipykernel_4460\2920196248.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Min-range product'

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df['price_range_loc'].value_counts(dropna = False)

Min-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [18]:
# Defining values unsing If-Statements with the loc() Function
df_ords_prods.loc[(df_ords_prods['prices'] <= 15) & (df_ords_prods['prices'] > 5), 'price_range_loc'] = 'Min-range product'

df_ords_prods.loc[df_ords_prods['prices'] > 15, 'price_range_loc'] = 'High-range product'

df_ords_prods.loc[df_ords_prods['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [19]:
df_ords_prods['price_range_loc'].value_counts(dropna = False)

Min-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [20]:
df_ords_prods['orders_day_of_the_week'].value_counts(dropna = False)

Saturday     6204182
Sunday       5660230
Friday       4496490
Monday       4213830
Thursday     4205791
Tuesday      3840534
Wednesday    3783802
Name: orders_day_of_the_week, dtype: int64

***Busiest day of the week: Saturday.
Slowest day of the week: Sunday***

### If-Statements with For-Loops

In [21]:
# Categorize the day of the week depending of the number of orders using For_loop
import time
start_time = time.time()

result = []

for value in df_ords_prods["orders_day_of_the_week"]:
  if value == 'Saturday':
    result.append("Busiest day")
  elif value == 'Wednesday':
    result.append("Least busy")
  else:
    result.append("Regularly busy")
print("--- %s seconds ---" % (time.time() - start_time))

--- 4.960312128067017 seconds ---


In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [23]:
df_ords_prods['busiest_day'] = result

In [24]:
# Check value frecuency in "busiest_day" 
df_ords_prods['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# 02. Create 'busiest_days' column

In [25]:
# Create a new colums. If-Statements with For_Loops. “Busiest days” (plural).
# Using '.time' function for fun.
start_time = time.time()

result2 = []

for value in df_ords_prods["orders_day_of_the_week"]:
  if value == 'Saturday' or value == 'Sunday':
    result2.append("Busiest days")
  elif value == 'Wednesday' or value =='Tuesday':
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")
print("--- %s seconds ---" % (time.time() - start_time))

--- 5.913800239562988 seconds ---


In [26]:
# Adding result2 to a new column "busiest_ days"
df_ords_prods['busiest_days'] = result2

In [27]:
df_ords_prods['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

In [28]:
df_ords_prods.shape

(32404859, 19)

***No Empty or unknown values in 'busiest_days' column.***

In [29]:
#Check the values of this new column for accuracy
df_ords_prods['busiest_days'].value_counts().describe()

count    3.000000e+00
mean     1.080162e+07
std      2.801404e+06
min      7.624336e+06
25%      9.744374e+06
50%      1.186441e+07
75%      1.239026e+07
max      1.291611e+07
Name: busiest_days, dtype: float64

***Observations: Total amount of orders within Slowestdays (Tuesday and Wednesday) 7624336 . 
              Total amount of orders within Busiest days (Saturday and Sunday) 1186864412. 
              Total amount of orders within Regularly busy days (Monday, Thursday and Friday) 12916111. 
              Average 10801620 orders among the three grouped days.
              Median 11864410 orders***

In [30]:
#Removing 'Unnamed' columns
df_ords_prods.drop(df_ords_prods.columns[df_ords_prods.columns.str.contains('unnamed',case = False)],axis = 1, inplace = True)

In [31]:
df_ords_prods.shape

(32404859, 17)

In [32]:
# Identify the busiest hours of the day.
df_ords_prods['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [33]:
# Periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.”
# Create a new column containing these labels called “busiest_period_of_day.”
result_per_hours = []

for value in df_ords_prods["order_hour_of_day"]:
  if value >= 9 and value <= 16:
    result_per_hours.append("Most orders")
  elif value == 23 or (value <= 6 and value >= 0):
    result_per_hours.append("Fewest orders")
  else:
    result_per_hours.append("Average orders")

In [34]:
# Creating a new column "busiest_period_of_day"
df_ords_prods['busiest_period_of_day'] = result_per_hours

In [35]:
df_ords_prods.head()

order_id  user_id  order_number orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                 Monday                  8   
1   2398795        1             2                Tuesday                  7   
2    473747        1             3                Tuesday                 12   
3   2254736        1             4              Wednesday                  7   
4    431534        1             5              Wednesday                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Min-range product   
1         Soda        77              7     9.0  Min-range product   
2         Soda        77              7     9.0  Min-range product   
3         Soda        77              7     9.0  Min-range product   
4         Soda        77              7     9.0  Min-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

In [36]:
# Value Frequency in 'busiest_period_of_day'
df_ords_prods['busiest_period_of_day'].value_counts()

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

### Bonus task

***For a more accurate and effective method to prevent system failure, I suggest addressing the issue by identifying the busiest hours in the week, given the fact that most active hours could differ between the days of the week ***

In [37]:
# Sample of the busiest 42 hours in the week
df_ords_prods[['order_hour_of_day', 'orders_day_of_the_week']].value_counts(dropna = False).nlargest(42)

order_hour_of_day  orders_day_of_the_week
14                 Saturday                  564393
13                 Saturday                  557922
15                 Saturday                  551565
10                 Sunday                    549633
11                 Saturday                  535299
12                 Saturday                  534854
10                 Saturday                  515688
11                 Sunday                    509186
9                  Sunday                    505499
16                 Saturday                  501181
12                 Sunday                    462540
13                 Sunday                    457030
14                 Sunday                    452772
15                 Sunday                    442473
9                  Saturday                  440896
16                 Sunday                    418173
17                 Saturday                  395353
14                 Friday                    390749
13                 Fri

***While 'Thursday' is categorized as a day with "Fewer orders," in some hours appear to be among the busiest time between 12:00 and 16:00***

In [39]:
# Export data frame in a suitable format, keep in mind the size of the data
start_time = time.time()

df_ords_prods.to_pickle(os.path.join(path_prepared, 'df_ords_prods_4.7_.pkl'))

print("--- %s seconds ---" % (time.time() - start_time))

--- 7.979761123657227 seconds ---
